### DISCUSSION TITLE
queue_declare does not receive the callback at random times

### DISCUSSION THREAD

**Discussion by kac723** [pika/pika#1481] [Q&A]
Hi,

I'm experiencing an issue with the `queue_declare` function where it won't receive the callback after multiple calls. The issue is random as it can happen on the 10th, 100th or 1000th call to `queue_declare`. Once the issue occurs, the channel and connection cannot be closed properly as the connection/channel close functions also don't receive the callbacks. I'm attaching scripts for the reproduction of the issue: [test.zip](https://github.com/user-attachments/files/16469080/test.zip)

To run the test, please extract the file and run `python3 test.py`. It will print multiple logs but on test failure, you should see: `Test failed. getQueueSize took more than 10 seconds. This will cause the script to block as we cannot close consumers`. All channels and connections will then be closed but one of them will likely hang. The hanging queue can be deduced from logs:
`Shutting down connection and channel for queue test0`
and the following log will be missing:
`RMQ shutdown - connection closed for queue test0`

Any help will be greatly appreciated.

Environment:
Python: 3.10.12
Erlang: 1:25.3.2.2-1
Pika: 1.2.0-1

**Comment by lukebakken** [comment]
I have added your code here, with formatting fixes - https://github.com/lukebakken/pika-1480

What does RabbitMQ log during your test run? Please compress the FULL RabbitMQ log file, and _ATTACH_ it to this issue.

One thing I notice right away is the use of `time.sleep` in the thread running Pika. This WILL block Pika's I/O loop. I don't remember the correct way to sleep when using `SelectConnection` off the top of my head, but I'm sure you could figure it out by looking at Pika's source.

In addition, your use of `event.wait` will block the thread on which it's running, which happens to be Pika's I/O thread. This is most likely the reason you are seeing this issue. You should use [`queue`](https://docs.python.org/3/library/queue.html) to synchronize between the thread running Pika's I/O loop and your consumer.

Please fork my repository, and make your changes based on that. I won't make them for you, but I can review them.

Thanks

**Comment by shinji-s** [comment]
Another issue is that channels are created in RMQ threads and the main thread calls 'queue_declare' on the channels in RMQ.getQueueSize() if I read the code correctly. AFAIK Pika connections/channels can't be shared between threads.

**Comment by kac723** [comment]
@shinji-s , could you please advise how should I request the queue size on a thread that's an RMQ thread rather than from main one?

**Comment by kac723** [comment]
I've pushed changes to my fork to replace time.sleep with Queue synchronisation but the issue remains.

**Comment by kac723** [comment]
Attached RMQ log file
[rmqlog.zip](https://github.com/user-attachments/files/16472551/rmqlog.zip)

**Comment by lukebakken** [comment]
@kac723 - please, fork _my_ repository so I can see your changes.

Also, before I spend more time with your code, please explain what you are trying to accomplish. Just consume messages?

**Comment by kac723** [comment]
I would have thought that this is the correct fork: https://github.com/kac723/pika-1480 
My aim is to be able to have an RMQConsumer object on which I can call getQueueSize at 1s intervals to check if something appeared in the RMQ queue

**Comment by lukebakken** [comment]
Thanks, for whatever reason I couldn't easily find this link - https://github.com/lukebakken/pika-1480/forks

I will return to this question when I have time.
